docs/proposals/image-provenance.md

```diff
@@ -0,0 +1,320 @@
+ 	Spec ImageReviewSpec
+ 
+ 	// Status is filled in by the backend and indicates whether the pod should be allowed.
+ 	Status ImageReviewStatus
+ }
+ 
+// ImageReviewSpec is a description of the token authentication request.
+type ImageReviewSpec struct {
+	// Images are the names of all images used in the Pod (one per container, optionally de-duplicated).
+	Images []string
+	// Annotations is a list of key-value pairs extracted from the Pod's annotations.
+	// It only includes keys which match the pattern `*.image-policy.k8s.io/*`.
+	// It is up to each webhook backend to determine how to interpret these annotations, if at all.
+        Annotations map[string]string
+	// Namespace is the namespace the pod is being created in.
+	Namespace string
+}
+
+// ImageReviewStatus is the result of the token authentication request.
+type ImageReviewStatus struct {
+	// Allowed indicates that all images were allowed to be run.
+	Allowed bool
+	// Error should be empty unless Allowed is false in which case it
+        // may contain a short description of what is wrong.  Kubernetes
+        // may truncate excessively long errors when displaying to the user.
+	Error string
+}
+```
+
+## Extending with Annotations
+
+All annotations on a Pod that match `*.image-policy.k8s.io/*` are sent to the webhook.
+Sending annotations allows users who are aware of the image policy backend to send
+extra information to it, and for different backends implementations to accept
+different information.  
+
+Examples of information you might put here are
+
+- request to "break glass" to override a policy, in case of emergency.
+- a ticket number from a ticket system that documents the break-glass request
+- provide a hint to the policy server as to the imageID of the image being provided, to save it a lookup
+
+In any case, the annotations are provided by the user and are not validated by Kubernetes in any way.  In the future, if an annotation is determined to be widely
+useful, we may promote it to a named field of ImageReviewSpec.
+
+In the case of a Pod update, Kubernetes may send the backend either all images in the updated image, or only the ones that
+changed, at its discretion.
+
+## Interaction with Controllers
+
+In the case of a Deployment object, no image check is done when the Deployment object is created or updated.
+Likewise, no check happens when the Deployment controller creates a ReplicaSet.  The check only happens
+when the ReplicaSet controller creates a Pod.  Checking Pod is necessary since users can directly create pods,
+and since third-parties can write their own controllers, which kubernetes might not be aware of or even contain
+pod templates.
+
+The ReplicaSet, or other controller, is responsible for recognizing when a 403 has happened
+(whether due to user not having permission due to bad image, or some other permission reason)
+and throttling itself and surfacing the error in a way that CLIs and UIs can show to the user.
```

Comment:
> I'm starting to feel like we're not doing enough to actually solve the usability problem here - the approach you describe is correct (RS needs to surface that info) but we're doing a terrible job of it. For instance... should RS have a condition "CreationForbidden": "true" with a reason "ImageRejected"? Should the deployment also have a condition surfaced?

> Right now, we see this all the time in OpenShift (by virtue of security policy) and end users really suffer without this info being surfaced (in quota, in policy, and in crash looping pods).

Conditions are one thing that may or may not help here but I believe we already have a mechanism for reporting failures, albeit we are doing a bad job with it: events. Events are nice but fetching exactly those you are interested in is impossible currently. It feels like we should have events as a subresource to all objects or something. https://github.com/kubernetes/kubernetes/issues/11994 is related.